dev-lang/rust/rust-1.18.0.recipe

```diff
@@ -0,0 +1,217 @@
+CHECKSUM_SHA256_3="dc7240d60a869fa24a68c8734fb7c810c27cca0a6dad52df6279865e4e8e7fae"
+CHECKSUM_SHA256_4="623889a326754e93a4a7a8a3487699ea786f4ca9b118f7468c12f8b9966905e6"
+CHECKSUM_SHA256_5="3616106943b3d3a345fc85ec02e9e125f80f60ea2f9384d4c5d6286c7142cfbd"
+CHECKSUM_SHA256_6="a85cc4847370573d194af95d6794304479d1024863c754d9b4e2499478c41393"
+CHECKSUM_SHA256_7="9ea59d17a7fa81aa4bdefa2bb45a5315219414386753eaf0988cadd465550b40"
+CHECKSUM_SHA256_8="ee080d47da2204094195362d91818cd56350c9c7e0969d78fc6d55c59529fce4"
+CHECKSUM_SHA256_9="7fca934f5ab6c08ba38376a6847cf6c6a1405794148d107bffa715c46f634bb4"
+CHECKSUM_SHA256_10="31d77c9ccb2924f8381a09b6eb8b7efe74fbc280c581a21728e0c50878413ec0"
+CHECKSUM_SHA256_11="e1ad31337422f9b05f57f43266048935f246cc35ca53671feceb9e6208bc4b40"
+CHECKSUM_SHA256_12="927f6b782f0be32c6027335bbd49fb40cffa922095ffacc9a0aca78e45023f1a"
+SOURCE_DIR="rust-$srcGitRev"
+PATCHES="rust-$portVersion.patchset"
+PATCHES_3="rust-installer-$portVersion.patchset"
+PATCHES_7="cargo-$cargoVersion.patchset"
+PATCHES_11="rust-rls-$portVersion.patchset"
+PATCHES_12="cargo-vendor-$srcGitRev12.patchset"
+REVISION="1"
+ARCHITECTURES="!x86_gcc2 ?x86 x86_64"
+SECONDARY_ARCHITECTURES="x86"
+
+PROVIDES="
+	rust$secondaryArchSuffix = $portVersion
+	cmd:rustc = $portVersion
+	cmd:rustdoc = $portVersion
+	cmd:rust_gdb = $portVersion
+	cmd:rust_lldb = $portVersion
+	cmd:cargo$secondaryArchSuffix = $cargoVersion
+	cmd:rls$secondaryArchSuffix = $portVersion
+	"
+REQUIRES="
+	haiku$secondaryArchSuffix
+	lib:libcrypto$secondaryArchSuffix
+	lib:libcurl$secondaryArchSuffix
+	lib:libssl$secondaryArchSuffix
+	lib:libz$secondaryArchSuffix
+	"
+BUILD_REQUIRES="
+	haiku${secondaryArchSuffix}_devel
+	devel:libcurl$secondaryArchSuffix
+	devel:libssl$secondaryArchSuffix
+	devel:libz$secondaryArchSuffix
+	"
+BUILD_PREREQUIRES="
+	cmd:cargo$secondaryArchSuffix == 0.18.0
+	cmd:cmake
+	cmd:cmp
+	cmd:file
+	cmd:find
+	cmd:git
+	cmd:gcc$secondaryArchSuffix
+	cmd:grep
+	cmd:ld$secondaryArchSuffix
+	cmd:make
+	cmd:pkg_config$secondaryArchSuffix
+	cmd:python
+	cmd:rustc == 1.17.0
+	cmd:sed
+	cmd:tar
+	cmd:which
+	cmd:xargs
+	"
+
+BUILD()
+{
+	# With the new build system, Rust always makes a tarball with the source
+	# In older recipes we symlinked the other git submodules, now we have to
+	# copy them
+	rm -rf cargo src/compiler-rt src/llvm src/rust-installer \
+	       src/liblibc src/rt/hoedown src/doc/nomicon src/doc/reference \
+	       src/doc/book rls
+	cp -r /sources-2/compiler-rt-$srcGitRev2 src/compiler-rt
+	cp -r /sources-3/rust-installer-$srcGitRev3 src/rust-installer
+	cp -r /sources-4/hoedown-$srcGitRev4 src/rt/hoedown
+	cp -r /sources-5/libc-$srcGitRev5 src/liblibc
+	cp -r /sources-6/llvm-$srcGitRev6 src/llvm
+	cp -r /sources-7/cargo-$cargoVersion cargo
+	cp -r /sources-8/nomicon-$srcGitRev8 src/doc/nomicon
+	cp -r /sources-9/reference-$srcGitRev9 src/doc/reference
+	cp -r /sources-10/book-$srcGitRev10 src/doc/book
+	cp -r /sources-11/rls-$srcGitRev11 rls
```

Comment:
I'd recommend using `"$sourceDirX"/` instead of `/sources-X/` (with X going from 2 to 11):
```diff
-	cp -r /sources-2/compiler-rt-$srcGitRev2 src/compiler-rt
-	cp -r /sources-3/rust-installer-$srcGitRev3 src/rust-installer
-	cp -r /sources-4/hoedown-$srcGitRev4 src/rt/hoedown
-	cp -r /sources-5/libc-$srcGitRev5 src/liblibc
-	cp -r /sources-6/llvm-$srcGitRev6 src/llvm
-	cp -r /sources-7/cargo-$cargoVersion cargo
-	cp -r /sources-8/nomicon-$srcGitRev8 src/doc/nomicon
-	cp -r /sources-9/reference-$srcGitRev9 src/doc/reference
-	cp -r /sources-10/book-$srcGitRev10 src/doc/book
-	cp -r /sources-11/rls-$srcGitRev11 rls
+	ln -sf "$sourceDir2"/compiler-rt-$srcGitRev2 src/compiler-rt
+	ln -sf "$sourceDir3"/rust-installer-$srcGitRev3 src/rust-installer
+	ln -sf "$sourceDir4"/hoedown-$srcGitRev4 src/rt/hoedown
+	ln -sf "$sourceDir5"/libc-$srcGitRev5 src/liblibc
+	ln -sf "$sourceDir6"/llvm-$srcGitRev6 src/llvm
+	ln -sf "$sourceDir7"/cargo-$cargoVersion cargo
+	ln -sf "$sourceDir8"/nomicon-$srcGitRev8 src/doc/nomicon
+	ln -sf "$sourceDir9"/reference-$srcGitRev9 src/doc/reference
+	ln -sf "$sourceDir10"/book-$srcGitRev10 src/doc/book
+	ln -sf "$sourceDir11"/rls-$srcGitRev11 rls
```
Changing each `/sources-X/` into `"$sourceDirX"/` is prefered.

And using `ln -sf` (or `ln -fs` ;) instead of `cp -r` is recommended if the build accepts this.
I think this might be an improvement because it has a small advantage in case the build is kicked, then interrupted before completion, then kicked again without first cleaning the work directory.
Indeed, in such a case, when hailuporter is run the second time, each destination would otherwise have 2 copies of its sources-X directory, the second copy landing _inside_ the first one.

**IMPORTANT:** I did not try this yet, so I don't know if the build will be fine with symbolic links instead of real directory trees.
So, if it works for this recipe (`0.18.0`), you might wish to also apply the same suggestion for `0.19.0` and maybe `0.17.0` too.

Hmm, maybe you and/or some other contributors prefer to keep this unchanged, so please feel free to wait for comments about this suggestion.

Reply:
In case the build does not work with symbolic links, there's something we might be able to do: create a `PATCH()` function and move the bloc of `cp -r` inside that function.

Hmm, this might slow down the process if haikuporter is not called with the `--no-git-repo` (or `-G`) switch. So I don't really know if this would be an improvement...

Reply:
I guess you missed the rm -r line before all the copying. I agree with you the copying is tedious, I build on a slow disk (USB 2) so this is the most annoying part of fixing and rebuilding the recipe

Reply:
I have tried to use symbolic links before... Something went wrong in those instances. However, the build system changed dramatically since Rust 1.15.0, so I will try again in the 1.20.0 recipe.

Reply:
> I guess you missed the rm -r line before all the copying.

True :)

BTW, in case the build does not modify any of the copies of the files from these `$sourceDirX` trees then using `cp --recursive --symbolic-link` (instead of just `cp --recursive`) might be a good idea to speed up this step.

Reply:
Also replying to myself here: the rust build system generates a source tarball, which breaks when symlinks are used. This is actually in the comment of the file. This is why we have comments.

In future recipes (1.20 and up) I will try to use symlinks again, as there now is an option not to build a src tarball. This might also get rid of the dependency on cargo-vendor.